state-chain/cf-integration-tests/src/fee_scaling.rs

```diff
@@ -0,0 +1,152 @@
+		size_change: IncreaseOrDecrease::Decrease(RangeOrderSize::Liquidity {
+			liquidity: 1_000u128,
+		}),
+	})
+}
+
+#[test]
+fn fee_scales_within_a_pool() {
+	const EPOCH_BLOCKS: u32 = 100;
+	const MAX_AUTHORITIES: AuthorityCount = 10;
+	let lp = AccountKeyring::Alice;
+	super::genesis::with_test_defaults()
+		.epoch_duration(EPOCH_BLOCKS)
+		.max_authorities(MAX_AUTHORITIES)
+		.with_additional_accounts(&[(
+			lp.to_account_id(),
+			AccountRole::LiquidityProvider,
+			5 * FLIPPERINOS_PER_FLIP,
+		)])
+		.build()
+		.execute_with(|| {
+			let (mut testnet, _, _) =
+				crate::network::fund_authorities_and_join_auction(MAX_AUTHORITIES);
+
+			let lp_account_id = lp.to_account_id();
+
+			let call = update_range_order_call(Asset::Eth);
+
+			let (_call_info_id, upfront_fee) = <Runtime as pallet_cf_flip::Config>::FeeScalingCallInfoIdentifier::call_info_and_spam_prevention_upfront_fee(
+				&call,
+				&lp_account_id,
+			).expect("We chose a call that should be scaled.");
+
+			assert_eq!(FeeScalingRate::<Runtime>::get(), FeeScalingRateConfig::NoScaling);
+
+			let (mut last_gas, mut last_remaining_balance) =
+				apply_extrinsic_and_calculate_gas_fee(lp, call.clone()).unwrap();
+			(0..50).for_each(|_| {
+				let (gas, remaining_balance) =
+					apply_extrinsic_and_calculate_gas_fee(lp, call.clone()).unwrap();
+				assert_ne!(gas, 0);
+				assert_eq!(gas, last_gas);
+				last_gas = gas;
+				assert!(remaining_balance < last_remaining_balance);
+				last_remaining_balance = remaining_balance;
+			});
+
+			// Reset the fee scaling counters
+			testnet.move_forward_blocks(1);
+
+			// Set the config to scale per pool.
+			FeeScalingRate::<Runtime>::set(FeeScalingRateConfig::ExponentBuffer {
+				buffer: 0,
+				exp_base: FixedU64::from_rational(2, 1u128),
+			});
+
+			let (mut last_gas, mut last_remaining_balance) =
+				apply_extrinsic_and_calculate_gas_fee(lp, call.clone()).unwrap();
+
+			let mut can_pay = true;
+			(1..50).for_each(|_| {
+				match apply_extrinsic_and_calculate_gas_fee(lp, call.clone()) {
+					Ok((gas, remaining_balance)) => {
+						// Either the gas is increasing, or we've hit the ceiling of the upfront fee.
+						assert!(gas > last_gas || gas == upfront_fee);
+						assert_eq!(remaining_balance, last_remaining_balance - gas);
+						// We should never fail, and then succeed again, since the fee always
+						// increases within a block in the same pool.
+						assert!(can_pay);
+						last_remaining_balance = remaining_balance;
+						last_gas = gas;
+					},
+					Err(e) => {
+						can_pay = false;
+						// We no longer have enough to pay fees.
+						assert_eq!(
+							e,
+							TransactionValidityError::Invalid(InvalidTransaction::Payment)
+						);
+						// No balance change if transaction fails
+						assert_eq!(Flip::total_balance_of(&lp_account_id), last_remaining_balance);
+					},
+				}
+			});
+
+			// We should have run out of balance by the end.
+			assert!(!can_pay);
```

Comment:
I don't think this is a necessary test condition. It also means that the test could succeed eg. because fees are applied incorrectly and immediately deduct the full balance.